File: NextApp/iOS/OrgPortalNext/Feature/Tools/VideoQuestionCsvExporter.swift
import Foundation
import Model

public enum VideoQuestionCsvExporter {
    public static func data(for questions: [VideoQuestion], video: DistributedVideo) -> Data {
        data(for: questions, videosByID: [video.id: video], fallbackVideo: video)
    }

    public static func data(for questions: [VideoQuestion], videos: [DistributedVideo]) -> Data {
        data(
            for: questions,
            videosByID: Dictionary(uniqueKeysWithValues: videos.map { ($0.id, $0) }),
            fallbackVideo: nil
        )
    }

    private static func data(
        for questions: [VideoQuestion],
        videosByID: [String: DistributedVideo],
        fallbackVideo: DistributedVideo?
    ) -> Data {
        let formatter = ISO8601DateFormatter()
        let lines = [
            "schema_version=1",
            "video_title,video_url,playback_seconds,question,answer,status,created_at,answered_at,memo",
        ] + questions.map { question in
            let video = videosByID[question.videoId] ?? fallbackVideo
            return [
                question.videoTitle.isEmpty ? (video?.title ?? "") : question.videoTitle,
                video.map(exportedVideoUrl) ?? "",
                String(question.playbackSeconds),
                question.questionText,
                question.answerText,
                question.isAnswered ? "answered" : "unanswered",
                question.createdAt.map { formatter.string(from: $0) } ?? "",
                question.answeredAt.map { formatter.string(from: $0) } ?? "",
                question.memoText,
            ]
            .map(escape)
            .joined(separator: ",")
        }

        let csv = lines.joined(separator: "\r\n") + "\r\n"
        return Data([0xEF, 0xBB, 0xBF]) + Data(csv.utf8)
    }

    private static func exportedVideoUrl(_ video: DistributedVideo) -> String {
        if !video.vimeoUrl.isEmpty { return video.vimeoUrl }
        return video.videoUrl
    }

    private static func escape(_ value: String) -> String {
        "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
    }
}

File: NextApp/Android/feature/tools/src/main/kotlin/jp/komehyappyo/member/next/feature/tools/VideoQuestionCsvExporter.kt
package jp.komehyappyo.member.next.feature.tools

import jp.komehyappyo.member.next.core.model.DistributedVideo
import jp.komehyappyo.member.next.core.model.VideoQuestion
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

object VideoQuestionCsvExporter {
    private val formatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME

    fun export(
        questions: List<VideoQuestion>,
        video: DistributedVideo,
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): String = exportRows(
        questions = questions,
        videosById = mapOf(video.id to video),
        fallbackVideo = video,
        zoneId = zoneId,
    )

    fun export(
        questions: List<VideoQuestion>,
        videos: List<DistributedVideo>,
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): String = exportRows(
        questions = questions,
        videosById = videos.associateBy { it.id },
        fallbackVideo = null,
        zoneId = zoneId,
    )

    private fun exportRows(
        questions: List<VideoQuestion>,
        videosById: Map<String, DistributedVideo>,
        fallbackVideo: DistributedVideo?,
        zoneId: ZoneId,
    ): String {
        val headerLines = listOf(
            "schema_version=1",
            "video_title,video_url,playback_seconds,question,answer,status,created_at,answered_at,memo",
        )
        val dataLines = questions.map { question ->
            val video = videosById[question.videoId] ?: fallbackVideo
            listOf(
                question.videoTitle.ifEmpty { video?.videoTitle.orEmpty() },
                video?.let(::exportedVideoUrl).orEmpty(),
                question.playbackSeconds.toString(),
                question.questionText,
                question.answerText,
                if (question.isAnswered) "answered" else "unanswered",
                question.createdAt?.let { timestamp ->
                    formatter.format(
                        Instant.parse(timestamp).atZone(zoneId),
                    )
                } ?: "",
                question.answeredAt?.let { timestamp ->
                    formatter.format(
                        Instant.parse(timestamp).atZone(zoneId),
                    )
                } ?: "",
                question.memoText,
            ).joinToString(",") { value -> '"' + value.replace("\"", "\"\"") + '"' }
        }
        return (headerLines + dataLines).joinToString("\r\n") + "\r\n"
    }

    private fun exportedVideoUrl(video: DistributedVideo): String {
        return video.vimeoUrl.ifEmpty { video.videoUrl }
    }
}
